Userland/Libraries/LibWeb/DOM/Node.cpp

```diff
@@ -470,7 +470,12 @@ void Node::insert_before(JS::NonnullGCPtr<Node> node, JS::GCPtr<Node> child, boo
         }
 
         // 6. Run assign slottables for a tree with node’s root.
-        assign_slottables_for_a_tree(node_to_insert->root());
+        // AD-HOC: The call to `assign_slottables_for_a_tree` iterates over the node's root's entire subtree. That
+        //         iteration does nothing if the root is not a shadow root (see `find_slottables`). This iteration can
+        //         be very expensive as the HTML parser inserts nodes, especially on sites with many elements. So we
+        //         skip it if we know it's going to be a no-op anyways.
+        if (auto& node_root = node_to_insert->root(); node_root.is_shadow_root())
+            assign_slottables_for_a_tree(node_root);
```

Comment:
Should we do this inside `assign_slottables_for_a_tree()` instead?
That way all of its client get to avoid the unnecessary traversal busywork.

Reply:
Yep that works too!